File: 3027/Solution.java
public class Solution {
    public static void main(String[] args){
        int[][] case1 = { {1,1}, {2,2}, {3,3} };
        int[][] case2 = { {6,2}, {4,4}, {2,6} };
        int[][] case3 = { {3,1}, {1,3}, {1,1} };

        System.out.println(numberOfPairs(case1));
        System.out.println(numberOfPairs(case2));
        System.out.println(numberOfPairs(case3));
    }

    public static int numberOfPairs(int[][] points){
        int pairs = 0;

        for(int i=0; i<points.length; i++){
            for(int j=0; j<points.length; j++){
                if(i == j){
                    continue;
                }

                if(points[i][0] <= points[j][0] && points[i][1] >= points[j][1] && isNotWithinPair(points, i, j)){
                    pairs++;
                }
            }
        }

        return pairs;
    }

    public static boolean isNotWithinPair(int[][] points, int i, int j){
        for(int x=0; x<points.length; x++){
            if(x == i || x == j){
                continue;
            }

            if(points[x][0] >= points[i][0] && points[x][0] <= points[j][0] && points[x][1] <= points[i][1] && points[x][1] >= points[j][1]){
                return false;
            }
        }

        return true;
    }
}

File: 3027/solution.rs
fn main() {
    let case1: Vec<Vec<i32>> = vec![
        vec![1,1],
        vec![2,2],
        vec![3,3]
    ];
    let case2: Vec<Vec<i32>> = vec![
        vec![6,2],
        vec![4,4],
        vec![2,6]
    ];
    let case3: Vec<Vec<i32>> = vec![
        vec![3,1],
        vec![1,3],
        vec![1,1]
    ];
    
    println!("Result: {}", number_of_pairs(&case1));
    println!("Result: {}", number_of_pairs(&case2));
    println!("Result: {}", number_of_pairs(&case3));
}

fn number_of_pairs(points: &Vec<Vec<i32>>) -> i32 {
    let mut pairs = 0;

    for i in 0..points.len() {
        for j in 0..points.len() {
            if i == j {
                continue;
            }

            if points[i][0] <= points[j][0] && points[i][1] >= points[j][1] && is_not_within_pair(&points, i, j){
                pairs+=1;
            }
        }
    }

    return pairs;
}

fn is_not_within_pair(points: &Vec<Vec<i32>>, i: usize, j: usize) -> bool {
    for x in 0..points.len() {
        if x == i || x == j {
            continue;
        }

        if points[x][0] >= points[i][0] && points[x][0] <= points[j][0] && points[x][1] <= points[i][1] && points[x][1] >= points[j][1]{
            return false;
        }
    }

    return true;
}
